diff --git a/drivers/vservices/session.c b/drivers/vservices/session.c
--- a/drivers/vservices/session.c
+++ b/drivers/vservices/session.c
@@ -1905,12 +1905,9 @@ int vs_session_handle_message(struct vs_session_device *session,
 		struct vs_mbuf *mbuf, vs_service_id_t service_id)
 {
 	struct vs_service_device *service;
-	struct vs_transport *transport;
 	unsigned long flags;
 
-	transport = session->transport;
-
 	service = vs_session_get_service(session, service_id);
 	if (!service) {
 		dev_err(&session->dev, "message for unknown service %d\n",
 				service_id);
