diff --git a/drivers/input/touchscreen/st/fts_lib/ftsIO.c b/drivers/input/touchscreen/st/fts_lib/ftsIO.c
--- a/drivers/input/touchscreen/st/fts_lib/ftsIO.c
+++ b/drivers/input/touchscreen/st/fts_lib/ftsIO.c
@@ -81,15 +81,15 @@ int openChannel(struct i2c_client *clt)
 	return OK;
 }
 
-struct device *getDev()
+struct device *getDev(void)
 {
 	if (client != NULL)
 		return &(client->dev);
 	else
 		return NULL;
 }
 
-struct i2c_client *getClient()
+struct i2c_client *getClient(void)
 {
 	if (client != NULL)
 		return client;
diff --git a/drivers/input/touchscreen/st/fts_lib/ftsTime.c b/drivers/input/touchscreen/st/fts_lib/ftsTime.c
--- a/drivers/input/touchscreen/st/fts_lib/ftsTime.c
+++ b/drivers/input/touchscreen/st/fts_lib/ftsTime.c
@@ -90,7 +90,7 @@ int elapsedNanosecond(struct StopWatch *w)
 	return result;
 }
 
-char *timestamp()
+char *timestamp(void)
 {
 	char *result = NULL;
 
