diff --git a/arch/arm64/kernel/Makefile b/arch/arm64/kernel/Makefile
index e1df5af..43b1339 100644
--- a/arch/arm64/kernel/Makefile
+++ b/arch/arm64/kernel/Makefile
@@ -42,10 +42,6 @@ arm64-obj-$(CONFIG_ARM64_MODULE_PLTS)	+= module-plts.o
 arm64-obj-$(CONFIG_PERF_EVENTS)		+= perf_regs.o perf_callchain.o
 arm64-obj-$(CONFIG_HW_PERF_EVENTS)	+= perf_event.o
 
-ifeq ($(CONFIG_TRACING),y)
-arm64-obj-$(CONFIG_HW_PERF_EVENTS)	+= perf_trace_counters.o perf_trace_user.o
-endif
-
 arm64-obj-$(CONFIG_HAVE_HW_BREAKPOINT)	+= hw_breakpoint.o
 arm64-obj-$(CONFIG_CPU_PM)		+= sleep.o suspend.o
 arm64-obj-$(CONFIG_CPU_IDLE)		+= cpuidle.o
diff --git a/kernel/sched/tune.c b/kernel/sched/tune.c
index b04122d..6c62069 100644
--- a/kernel/sched/tune.c
+++ b/kernel/sched/tune.c
@@ -575,6 +575,26 @@ int schedtune_prefer_idle(struct task_struct *p)
 	return prefer_idle;
 }
 
+bool schedtune_task_colocated(struct task_struct *p)
+{
+#ifdef CONFIG_SCHED_WALT
+	struct schedtune *st;
+	bool colocate;
+
+	if (unlikely(!schedtune_initialized))
+		return false;
+
+	rcu_read_lock();
+	st = task_schedtune(p);
+	colocate = st->colocate;
+	rcu_read_unlock();
+
+	return colocate;
+#else
+	return false;
+#endif
+}
+
 static u64
 prefer_idle_read(struct cgroup_subsys_state *css, struct cftype *cft)
 {
diff --git a/kernel/sched/tune.h b/kernel/sched/tune.h
index ec5e2bc..3a97f99 100644
--- a/kernel/sched/tune.h
+++ b/kernel/sched/tune.h
@@ -16,6 +16,7 @@ int schedtune_cpu_boost(int cpu);
 int schedtune_task_boost(struct task_struct *tsk);
 
 int schedtune_prefer_idle(struct task_struct *tsk);
+bool schedtune_task_colocated(struct task_struct *tsk);
 
 void schedtune_enqueue_task(struct task_struct *p, int cpu);
 void schedtune_dequeue_task(struct task_struct *p, int cpu);
@@ -35,7 +36,11 @@ void schedtune_dequeue_task(struct task_struct *p, int cpu);
 #define schedtune_prefer_idle(tsk) 0
 #endif
 
+#define schedtune_task_colocated(tsk) false
 #define schedtune_enqueue_task(task, cpu) do { } while (0)
 #define schedtune_dequeue_task(task, cpu) do { } while (0)
 
 #endif /* CONFIG_SCHED_TUNE */
+
+/* Prefer faster CPUs for tasks with a positive boost. */
+#define schedtune_prefer_high_cap(tsk) (schedtune_task_boost(tsk) > 0)
diff --git a/kernel/sched/walt.c b/kernel/sched/walt.c
index d6bf94a..3fd5d71 100644
--- a/kernel/sched/walt.c
+++ b/kernel/sched/walt.c
@@ -26,6 +26,7 @@
 #include <linux/jiffies.h>
 #include <linux/sched/stat.h>
 #include "sched.h"
+#include "tune.h"
 #include "walt.h"
 
 const char *task_event_names[] = {"PUT_PREV_TASK", "PICK_NEXT_TASK",
diff --git a/drivers/cpufreq/cpufreq_times.c b/drivers/cpufreq/cpufreq_times.c
index 8d9cc58..c4519b0 100644
--- a/drivers/cpufreq/cpufreq_times.c
+++ b/drivers/cpufreq/cpufreq_times.c
@@ -31,10 +31,6 @@ static DECLARE_HASHTABLE(uid_hash_table, UID_HASH_BITS);
 
 /* task->time_in_state */
 static  __cacheline_aligned_in_smp DEFINE_SPINLOCK(task_time_in_state_lock);
-/* task->concurrent_active_time */
-static  __cacheline_aligned_in_smp DEFINE_SPINLOCK(task_concurrent_active_time_lock);
-/* task->concurrent_policy_time */
-static  __cacheline_aligned_in_smp DEFINE_SPINLOCK(task_concurrent_policy_time_lock);
 /* uid_hash_table */
 static __cacheline_aligned_in_smp DEFINE_SPINLOCK(uid_lock);
 
@@ -327,14 +323,6 @@ void cpufreq_task_times_init(struct task_struct *p)
 	p->time_in_state = temp;
 	spin_unlock_irqrestore(&task_time_in_state_lock, flags);
 	p->max_state = max_state;
-
-	spin_lock_irqsave(&task_concurrent_active_time_lock, flags);
-	p->concurrent_active_time = NULL;
-	spin_unlock_irqrestore(&task_concurrent_active_time_lock, flags);
-
-	spin_lock_irqsave(&task_concurrent_policy_time_lock, flags);
-	p->concurrent_policy_time = NULL;
-	spin_unlock_irqrestore(&task_concurrent_policy_time_lock, flags);
 }
 
 /* Caller must hold task_time_in_state_lock */
diff --git a/include/linux/cgroup_subsys.h b/include/linux/cgroup_subsys.h
index 7366662..d99e801 100644
--- a/include/linux/cgroup_subsys.h
+++ b/include/linux/cgroup_subsys.h
@@ -21,7 +21,7 @@ SUBSYS(cpu)
 SUBSYS(cpuacct)
 #endif
 
-#if IS_ENABLED(CONFIG_CGROUP_SCHEDTUNE) || IS_ENABLED(CONFIG_SCHED_TUNE_DUMMY)
+#if IS_ENABLED(CONFIG_SCHED_TUNE) || IS_ENABLED(CONFIG_SCHED_TUNE_DUMMY)
 SUBSYS(schedtune)
 #endif
 
diff --git a/include/linux/sched/core_ctl.h b/include/linux/sched/core_ctl.h
index 281b852..ca4f2dd 100644
--- a/include/linux/sched/core_ctl.h
+++ b/include/linux/sched/core_ctl.h
@@ -14,6 +14,8 @@
 #ifndef __CORE_CTL_H
 #define __CORE_CTL_H
 
+#include <linux/sched/sysctl.h>
+
 struct core_ctl_notif_data {
 	unsigned int nr_big;
 	unsigned int coloc_load_pct;
diff --git a/kernel/sched/core_ctl.c b/kernel/sched/core_ctl.c
index c415761..1623ca3 100644
--- a/kernel/sched/core_ctl.c
+++ b/kernel/sched/core_ctl.c
@@ -20,6 +20,9 @@
 #include "sched.h"
 #include "walt.h"
 
+#define MAX_CPUS_PER_CLUSTER 6
+#define MAX_NR_HISTORY_COUNT	2
+
 struct cluster_data {
 	bool inited;
 	unsigned int min_cpus;
