diff --git a/fs/exec.c b/fs/exec.c
index e6644f3b3..679a388dc 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1965,6 +1965,11 @@ static int do_execveat_common(int fd, struct filename *filename,
 	putname(filename);
 	return retval;
 }
+#ifdef CONFIG_KSU_MANUAL_HOOK
+__attribute__((hot))
+extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr,
+				void *argv, void *envp, int *flags);
+#endif
 
 int kernel_execve(const char *kernel_filename,
 		  const char *const *argv, const char *const *envp)
@@ -2031,6 +2036,9 @@ static int do_execve(struct filename *filename,
 {
 	struct user_arg_ptr argv = { .ptr.native = __argv };
 	struct user_arg_ptr envp = { .ptr.native = __envp };
+	#ifdef CONFIG_KSU_MANUAL_HOOK
+	ksu_handle_execveat((int *)AT_FDCWD, &filename, &argv, &envp, 0);
+	#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
 
@@ -2074,6 +2082,9 @@ static int compat_do_execveat(int fd, struct filename *filename,
 		.is_compat = true,
 		.ptr.compat = __envp,
 	};
+	#ifdef CONFIG_KSU_MANUAL_HOOK // 32-bit ksud and 32-on-64 support
+	ksu_handle_execveat((int *)AT_FDCWD, &filename, &argv, &envp, 0);
+	#endif
 	return do_execveat_common(fd, filename, argv, envp, flags);
 }
 #endif
diff --git a/fs/open.c b/fs/open.c
index 6f1458976..a80c50678 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -489,9 +489,17 @@ static long do_faccessat(int dfd, const char __user *filename, int mode, int fla
 
 	return res;
 }
+#ifdef CONFIG_KSU_MANUAL_HOOK
+__attribute__((hot))
+extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user,
+				int *mode, int *flags);
+#endif
 
 SYSCALL_DEFINE3(faccessat, int, dfd, const char __user *, filename, int, mode)
 {
+	#ifdef CONFIG_KSU_MANUAL_HOOK
+		ksu_handle_faccessat(&dfd, &filename, &mode, NULL);
+	#endif
 	return do_faccessat(dfd, filename, mode, 0);
 }
 
diff --git a/fs/stat.c b/fs/stat.c
index ef50573c7..58edec56a 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -426,6 +426,16 @@ SYSCALL_DEFINE2(newlstat, const char __user *, filename,
 
 	return cp_new_stat(&stat, statbuf);
 }
+#ifdef CONFIG_KSU_MANUAL_HOOK
+__attribute__((hot))
+extern int ksu_handle_stat(int *dfd, const char __user **filename_user,
+				int *flags);
+
+extern void ksu_handle_newfstat_ret(unsigned int *fd, struct stat __user **statbuf_ptr);
+#if defined(__ARCH_WANT_STAT64) || defined(__ARCH_WANT_COMPAT_STAT64)
+extern void ksu_handle_fstat64_ret(unsigned long *fd, struct stat64 __user **statbuf_ptr); // optional
+#endif
+#endif
 
 #if !defined(__ARCH_WANT_STAT64) || defined(__ARCH_WANT_SYS_NEWFSTATAT)
 SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
@@ -433,7 +443,9 @@ SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
 {
 	struct kstat stat;
 	int error;
-
+	#ifdef CONFIG_KSU_MANUAL_HOOK
+		ksu_handle_stat(&dfd, &filename, &flag);
+	#endif
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
@@ -448,7 +460,9 @@ SYSCALL_DEFINE2(newfstat, unsigned int, fd, struct stat __user *, statbuf)
 
 	if (!error)
 		error = cp_new_stat(&stat, statbuf);
-
+	#ifdef CONFIG_KSU_MANUAL_HOOK
+		ksu_handle_newfstat_ret(&fd, &statbuf);
+	#endif
 	return error;
 }
 #endif
@@ -575,7 +589,9 @@ SYSCALL_DEFINE2(fstat64, unsigned long, fd, struct stat64 __user *, statbuf)
 
 	if (!error)
 		error = cp_new_stat64(&stat, statbuf);
-
+	#ifdef CONFIG_KSU_MANUAL_HOOK // for 32-bit
+		ksu_handle_fstat64_ret(&fd, &statbuf);
+	#endif
 	return error;
 }
 
@@ -584,7 +600,9 @@ SYSCALL_DEFINE4(fstatat64, int, dfd, const char __user *, filename,
 {
 	struct kstat stat;
 	int error;
-
+	#ifdef CONFIG_KSU_MANUAL_HOOK // 32-bit su
+		ksu_handle_stat(&dfd, &filename, &flag);
+	#endif
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
diff --git a/kernel/reboot.c b/kernel/reboot.c
index 182a2106c..d451d1d5f 100644
--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -699,13 +699,20 @@ DEFINE_MUTEX(system_transition_mutex);
  *
  * reboot doesn't sync: do that yourself before calling this.
  */
+
+#ifdef CONFIG_KSU_MANUAL_HOOK
+extern int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg);
+#endif
+
 SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 		void __user *, arg)
 {
 	struct pid_namespace *pid_ns = task_active_pid_ns(current);
 	char buffer[256];
 	int ret = 0;
-
+	#ifdef CONFIG_KSU_MANUAL_HOOK
+		ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
+	#endif
 	/* We only trust the superuser with rebooting the system. */
 	if (!ns_capable(pid_ns->user_ns, CAP_SYS_BOOT))
 		return -EPERM;
